### PR TITLE
Site Title, Post Title: Fix typography for blocks with `a` children

### PR DESCRIPTION
**Pull request by rafaelgallani**
Related to:
- #42404
- #41725
- #46180
- https://github.com/Automattic/wp-calypso/issues/92525

## What?
This PR fixes the typography properties for Site Title and Post Title blocks, enabling their specific styles to be respected/take precedence over more generic styles.

## Why?
Individual typography styles for the block are not respected if it acts as a link (if the "Make title a link" setting is enabled). Styles-defined settings are being used for these blocks. The text decoration, font family, and a few other attributes (described below) can't be set individually for these blocks, then.

## How?
- The `selectors` setting for the `typography` within the `block.json` of the blocks needs to be updated - as in the PR - to ensure any `theme.json` style change also affects the inner `a` element;
- There are also CSS rules to pass through editor-defined styles from the parent element to the `a` children, if present; otherwise, use whatever gets defined in Styles/`theme.json`.

## Testing Instructions
- To test block-specific styles:
  - Go to Appearance > Editor;
  - Select "Styles" and edit link-specific styles - set a different font family for them;
  - Then, in any template, add the "Site Title" block;
  - Try to change the font to one different than previously selected, and ensure it works properly.
- To test `theme.json` styles:
  - Activate `emptytheme`;
  - Define specific styles for the `core/site-title` element:
  - ```json
    "styles": {
        "blocks": {
            "core/site-title": {
                "typography": {
                    "fontFamily": "Comic Sans MS, sans-serif"
                }
            }
        }
    }
    ```
   - Go to Appearance > Editor;
   - Add the "Site Title" block in any template and ensure the font used in `theme.json` is being correctly displayed;
   - Change the font family for the block and ensure it gets correctly updated.

Any other typography property can also be used (font was just used as an example):
- Font;
- Size
- Appearance;
- Line height;
- Letter spacing;
- Decoration;
- Letter case.

## Screenshots or screencast <!-- if applicable -->
### Before
https://github.com/user-attachments/assets/937b6620-1a90-4b16-a7c0-4db7f4a81e80

### After
https://github.com/user-attachments/assets/e908141b-d770-41de-aeb2-7f081690855c

### CODE REVIEW

File: packages/block-library/src/post-title/style.scss
Comment (jorgefilipecosta) [inline packages/block-library/src/post-title/style.scss:6]: If we add more styling options we need to create another selector following the same logic. Plus these selectors don't look ideal and we are sending many bytes over the wise but it seems a fair trade-off.

**Review by jorgefilipecosta — COMMENTED** [review]
Thank you for working on this @rafaelgallani 👍
This solution works and personally, I'm ok with it, it is not ideal but there is not ideal solution in this case.

I would love a second option on this one to know if we can merge it. Any thoughts on this @noisysocks, @aaronrobertshaw, or @oandregal ?

**Comment by annezazu** [comment]
Noting that I tested this PR and confirmed it resolves https://github.com/WordPress/gutenberg/issues/41725

**Review by andrewserong — APPROVED** [review]
Nice work here!

> This solution works and personally, I'm ok with it, it is not ideal but there is not ideal solution in this case.
> I would love a second option on this one to know if we can merge it.

> If we add more styling options we need to create another selector following the same logic. Plus these selectors don't look ideal and we are sending many bytes over the wise but it seems a fair trade-off.

Good points. TL;DR: On balance I think I lean toward the idea of including the additional CSS here to factor in the link state, as it's a pretty common one.

---

Longer thoughts:

That said, these are both dynamic blocks where their markup is generated at render time. This means that it could be _technically_ possible to inject the desired inline styles into the `a` tag at render time if need be?

The main problem with doing that is that it would add further complexity to the PHP side of things. For example if we intercepted the attributes from [the call to `get_block_wrapper_attributes`](https://github.com/WordPress/gutenberg/blob/690d933cfb88a857521fd33fafd5a9e46aefb2f8/packages/block-library/src/post-title/index.php#L51) we could (theoretically) either grab the styles we need, or check for the presence of certain inline styles and then dynamically output the inherit rules as needed. That gets pretty messy once other styles like padding are involved, as we wouldn't want them to be duplicated on the `a` element. (Example wrapper attributes: `style="background-color:#f7a1a1; font-size:clamp(0.984rem, 0.984rem + ((1vw - 0.2rem) * 0.86), 1.5rem);line-height:1.1; border-color:#d27373;border-width:15px; padding-top:15px;padding-bottom:15px;padding-left:15px;padding-right:15px;" class="wp-block-post-title has-background has-border-color"`)

So, a second option (still PHP) could be to skip serialization for the Typography support, and call the relevant Typography and/or style engine functions from within the PHP version of these blocks, and then inject the results into the `a` tag instead of the wrapper. This again has the problem of added complexity, and would need to be duplicated in JS, too.

All up, this leans me to think that the approach in this PR is likely the simplest as it only adds some additional rules. My vote would be to go with the current state of this PR, but I mostly wanted to share the thoughts above as a potential alternative in case anyone else thought it worth pursuing.

Testing well for me:

✅ Font-weight at the individual block level overrides anchor styles
✅ Font-family at the individual block level overrides anchor styles
✅ Font-size at the individual block level overrides anchor styles
✅ Line-height at the individual block level overrides anchor styles
✅ Font-style at the individual block level overrides anchor styles
✅ Letter-spacing at the individual block level works as expected (I don't know if you can set it for the anchor in global styles, so I wasn't _too_ sure if the extra rule is needed, but it's working well in testing)
✅ Text-decoration at the individual block level works as expected. I noticed that it didn't always work reliably in global styles, but I don't think that's really related to this PR and not a blocker for proceeding.

LGTM! 🎉

It might be worth getting a second approving PR due to the discussion surrounding the CSS rules, but overall I'm in favour of this approach 👍

**Comment by rafaelgallani** [comment]
@andrewserong Wow, that's a very detailed overview. Thank you so much for that!

> So, a second option (still PHP) could be to skip serialization for the Typography support, and call the relevant Typography and/or style engine functions from within the PHP version of these blocks, and then inject the results into the a tag instead of the wrapper. This again has the problem of added complexity, and would need to be duplicated in JS, too.

I like this idea, but, yeah, I think we would need to duplicate this in JS. 
And I also think this would solve another related issue (which I'm still trying to find) of the heading styles (h1, h2...) not being applied to the `a` child tag.

Also, I just wanted to confirm before merging this one - it's been a while since my last contribution to Gutenberg: since we got a review from a `gutenberg-core` team member (you 😬) we're good to go, right? 
And.. Is there anything I need to add (a tag or comment) so my PR gets picked in the next RC? I'm basing my comment on [these guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/repository-management.md#merging-pull-requests).

**Comment by andrewserong** [comment]
> Also, I just wanted to confirm before merging this one - it's been a while since my last contribution to Gutenberg: since we got a review from a gutenberg-core team member (you 😬) we're good to go, right?

Yes, I think we're good to go. @jorgefilipecosta is also a `gutenberg-core` team member and seemed happy with this approach but wanted a second opinion, so with my review too and Anne confirming it fixes the issue, I think we're good here. We don't have a formal number of approvals process in the Gutenberg repo, so determining whether or not a PR is ready for merge is done on a bit of a case by case basis, depending on how risky a PR might seem (or how easy it might be to make further updates).

> And.. Is there anything I need to add (a tag or comment) so my PR gets picked in the next RC?

Nope! The next RC will be cut from `trunk` so merging a PR means it'll automatically get picked up in the next release cycle. If we're currently in an RC-phase (we're not just yet) then we can also backport urgent PRs to the current RC by flagging it for release leads via the `Backport to Gutenberg RC` label and manually updating the milestone after the PR is merged. But we don't have to do that for this one right now 🙂

This looks good to merge, to me!

**Comment by rafaelgallani** [comment]
> @jorgefilipecosta is also a `gutenberg-core` team member and seemed happy with this approach but wanted a second opinion, so with my review too and Anne confirming it fixes the issue, I think we're good here.

Yeah. I haven't noticed that he is also a team member, sorry 😅 

---

Alright then. Merging this one. Thanks!

**Comment by richtabor** [comment]
Noting an issue here:

From this theme.json: 

```
"core/site-title": {
                "elements": {
                    "link": {
                        "typography": {
                            "textDecoration": "none"
                        }
                    }
                },
                "typography": {
                    "fontSize": "var(--wp--preset--font-size--medium)",
                    "fontWeight": "550"
                }
            }
```

and 

```
"h1": {
                "typography": {
                    "fontSize": "var(--wp--preset--font-size--xx-large)",
                    "lineHeight": "1"
                }
            },
```

I'm getting this: 

CleanShot 2024-09-09 at 10 28 17

CleanShot 2024-09-09 at 10 26 45

Because the elements.h1 styles are applied to the `<h1>`, effectively applying a font size to the wrapping H1.

File: packages/block-library/src/site-title/block.json
Comment (aaronrobertshaw) [inline packages/block-library/src/site-title/block.json:80]: This introduces a regression where the editor and frontend apply different styles. The Site Title block doesn't render an inner `span` on the frontend.

As [proposed](https://github.com/WordPress/gutenberg/issues/65169#issuecomment-2345321749) over on https://github.com/WordPress/gutenberg/issues/65169, I have some ideas to address this and the other issue flagged in https://github.com/WordPress/gutenberg/pull/64911#issuecomment-2339018022.

I'll get a draft PR up soon.

**Pull request merged**
